libvast/vast/hash_index.hpp

```diff
@@ -0,0 +1,256 @@
+/******************************************************************************
+ *                    _   _____   __________                                  *
+ *                   | | / / _ | / __/_  __/     Visibility                   *
+ *                   | |/ / __ |_\ \  / /          Across                     *
+ *                   |___/_/ |_/___/ /_/       Space and Time                 *
+ *                                                                            *
+ * This file is part of VAST. It is subject to the license terms in the       *
+ * LICENSE file found in the top-level directory of this distribution and at  *
+ * http://vast.io/license. No part of VAST, including this file, may be       *
+ * copied, modified, propagated, or distributed except according to the terms *
+ * contained in the LICENSE file.                                             *
+ ******************************************************************************/
+
+#pragma once
+
+#include "vast/concept/hashable/uhash.hpp"
+#include "vast/concept/hashable/xxhash.hpp"
+#include "vast/data.hpp"
+#include "vast/detail/assert.hpp"
+#include "vast/detail/steady_map.hpp"
+#include "vast/detail/type_traits.hpp"
+#include "vast/value_index.hpp"
+#include "vast/view.hpp"
+
+#include <caf/deserializer.hpp>
+#include <caf/expected.hpp>
+#include <caf/optional.hpp>
+#include <caf/serializer.hpp>
+
+#include <algorithm>
+#include <array>
+#include <cstring>
+#include <string>
+#include <type_traits>
+#include <unordered_set>
+#include <vector>
+
+namespace vast {
+
+/// An index that only supports equality lookup by hashing its data. The
+/// hash_index computes a digest of the input data and concatenates all digests
+/// in a sequence. Optionally, it chops off the values after a fixed number of
+/// bytes for a more space-efficient representation, at the cost of more false
+/// positives. A separate "satellite structure" keeps track of hash collision
+/// to make the index exact. The additional state to build this satellite
+/// structure only exists during the construction of the index. Upon
+/// descruction, this extra state ceases to exist and it will not be possible
+/// to append further values when deserializing an existing index.
+template <size_t Bytes>
+class hash_index : public value_index {
+  static_assert(Bytes > 0);
+  static_assert(Bytes <= 8);
```

Comment:
```suggestion
  static_assert(Bytes <= sizeof(decltype(seeds_)::value_type);
```

Not a fan of magic numbers, especially since the size of a `size_t` may vary between platforms.

This additionally requires the assertion to be moved below the declaration of `seeds_`. Alternatively, you can always add a `using-declaration` above for the `size_t` that gives it a proper name.

Reply:
I should have commented the magic number. Here, I thought that it doesn't make sense keep digests larger than 8 bytes because they exceed the length of a 64-bit integer.

I'm relying on this in `key_hasher`. Probably adding a comment here is what would help?!

Reply:
> Here, I thought that it doesn't make sense keep digests larger than 8 bytes because they exceed the length of a 64-bit integer.

Not only that, but if `size_t` is less than 64-bit long on your platform, you start to get other issues. I'd make it `sizeof(size_t)` in both places, and maybe swap out `size_t` for `uint64_t` just to be sure.

Reply:
I've added more documentation. I think `size_t` is the right type for the `Bytes` parameter. It will never carry a hash and also serve in `std::array<T, Bytes>`. Hopefully the new documentation makes this clearer.

Reply:
It's not about the `size_t` for the template parameter `Bytes`, but rather about the value type  that you store the digests in.

Anyways, the comment does the job just as fine as the `sizeof`.

Reply:
Ah, I missed that. Thanks for clarifying.